mautic/mautic
# Rector resolving deprecations for Symfony 4.2

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Y]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [Y]
| Automated tests included?              | [Y, mostly existing already] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

This PR is increasing Rector automatic check from Symfony 4.1 o 4.2 depreactions. The main topic was to stop getting dependencies from container. I was able to remove some places and ignored other places for various reasons that are documented in the rector config file.

The biggest refactoring in this PR is related to removing the `LegacyLeadModel` as that was also flagged by the Rector rule. So I rather removed this deprecated class completely. It was quite a rabbit hole as it forced me to move some code around to avoid circular dependencies. Luckily, we have quite a lot of test covering the contact merging so I'm quite confident this did not break anything.

I added one more test to increase code coverage for DWC route. It turned out this route was broken for some time as it was linked to method that did not exist. Or the name was a bit different. So I fixed that route but I'm not quite sure if it's useful when no one missed it.

I had to use different approach to load fixtures during installations. Loading them from a folder did not work as we to put dependencies to them now when we cannot do it via the container. The replacement was quite elegant actually as we can use the `SymfonyFixturesLoader` instead and use already existing group to load the fixtures we want and use DI. However this service is not mockable so the existing unit test was failing because of it. I had to create a wrapper class with interface to fix this design flaw in Doctrine's service.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Try to merge contacts via API, import, UI, page hits
